sample_consensus/include/pcl/sample_consensus/impl/sac_model_sphere.hpp

```diff
@@ -366,24 +366,89 @@ pcl::SampleConsensusModelSphere<PointT>::optimizeModelCoefficients (
   {
     PCL_ERROR ("[pcl::SampleConsensusModelSphere::projectPoints] Given model is invalid!\n");
     return;
   }
 
-  // Allocate enough space and copy the basics
-  projected_points.resize (input_->size ());
   projected_points.header   = input_->header;
-  projected_points.width    = input_->width;
-  projected_points.height   = input_->height;
   projected_points.is_dense = input_->is_dense;
 
-  PCL_WARN ("[pcl::SampleConsensusModelSphere::projectPoints] Not implemented yet.\n");
-  projected_points.points = input_->points;
+  Eigen::Vector3f center (model_coefficients[0], model_coefficients[1], model_coefficients[2]);
+
+  // Copy all the data fields from the input cloud to the projected one?
+  if (copy_data_fields)
+  {
+    // Allocate enough space and copy the basics
+    projected_points.resize (input_->size ());
+    projected_points.width    = input_->width;
+    projected_points.height   = input_->height;
+
+    typedef typename pcl::traits::fieldList<PointT>::type FieldList;
+    // Iterate over each point
+    for (size_t i = 0; i < projected_points.points.size (); ++i)
+      // Iterate over each dimension
+      pcl::for_each_type <FieldList> (NdConcatenateFunctor <PointT, PointT> (input_->points[i], projected_points.points[i]));
+
+    // Iterate through the 3d points and calculate the distances from them to the sphere
+    for (size_t i = 0; i < inliers.size (); ++i)
+    {
+      // what i have:
+      // P : Sample Point
+      Eigen::Vector3d P (input_->points[inliers[i]].x, input_->points[inliers[i]].y, input_->points[inliers[i]].z);
+      // C : sphere Center
+      Eigen::Vector3d C (model_coefficients[0], model_coefficients[1], model_coefficients[2]);
+      // r : Radius
+      double r = model_coefficients[3];
+
+      Eigen::Vector3d helper_vectorPC = P - C;
+
+      // K : Point on Sphere
+      Eigen::Vector3d K = C + r * helper_vectorPC.normalized ();
+
+      projected_points.points[i].x = static_cast<float> (K[0]);
+      projected_points.points[i].y = static_cast<float> (K[1]);
+      projected_points.points[i].z = static_cast<float> (K[2]);
+    }
+  }
+  else
+  {
+    // Allocate enough space and copy the basics
+    projected_points.resize (inliers.size ());
+    projected_points.width    = static_cast<uint32_t> (inliers.size ());
+    projected_points.height   = 1;
+
+    typedef typename pcl::traits::fieldList<PointT>::type FieldList;
+    // Iterate over each point
+    for (size_t i = 0; i < inliers.size (); ++i)
+      // Iterate over each dimension
+      pcl::for_each_type <FieldList> (NdConcatenateFunctor <PointT, PointT> (input_->points[inliers[i]], projected_points.points[i]));
+
+    // Iterate through the 3d points and calculate the distances from them to the plane
+    for (size_t i = 0; i < inliers.size (); ++i)
+    {
+      // what i have:
+      // P : Sample Point
+      Eigen::Vector3d P (input_->points[inliers[i]].x, input_->points[inliers[i]].y, input_->points[inliers[i]].z);
```

Comment:
vector3fMap?

Reply:
Not easily usable because the rest is in double precision. A cast would be possible of course, but this line is is not so bad IMO.
I applied most of the other changes you suggested, feel free to make further improvements